// @langchain/langgraph/dist/pregel/algo.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._prepareNextTasks = exports._applyWrites = exports._localWrite = exports._localRead = exports.shouldInterrupt = exports.increment = void 0;
/* eslint-disable no-param-reassign */
const runnables_1 = require("@langchain/core/runnables");
const langgraph_checkpoint_1 = require("@langchain/langgraph-checkpoint");
const base_js_1 = require("../channels/base.cjs");
const io_js_1 = require("./io.cjs");
const constants_js_1 = require("../constants.cjs");
const errors_js_1 = require("../errors.cjs");
const utils_js_1 = require("./utils.cjs");
const increment = (current) => {
    return current !== undefined ? current + 1 : 1;
};
exports.increment = increment;
function shouldInterrupt(checkpoint, interruptNodes, tasks) {
    const versionValues = Object.values(checkpoint.channel_versions);
    const versionType = versionValues.length > 0 ? typeof versionValues[0] : undefined;
    let nullVersion;
    if (versionType === "number") {
        nullVersion = 0;
    }
    else if (versionType === "string") {
        nullVersion = "";
    }
    const seen = checkpoint.versions_seen[constants_js_1.INTERRUPT] ?? {};
    const anyChannelUpdated = Object.entries(checkpoint.channel_versions).some(([chan, version]) => {
        return version > (seen[chan] ?? nullVersion);
    });
    const anyTriggeredNodeInInterruptNodes = tasks.some((task) => interruptNodes === "*"
        ? !task.config?.tags?.includes(constants_js_1.TAG_HIDDEN)
        : interruptNodes.includes(task.name));
    return anyChannelUpdated && anyTriggeredNodeInInterruptNodes;
}
exports.shouldInterrupt = shouldInterrupt;
function _localRead(checkpoint, channels, task, select, fresh = false) {
    if (fresh) {
        const newCheckpoint = (0, base_js_1.createCheckpoint)(checkpoint, channels, -1);
        // create a new copy of channels
        const newChannels = (0, base_js_1.emptyChannels)(channels, newCheckpoint);
        // Note: _applyWrites contains side effects
        _applyWrites((0, langgraph_checkpoint_1.copyCheckpoint)(newCheckpoint), newChannels, [task]);
        return (0, io_js_1.readChannels)(newChannels, select);
    }
    else {
        return (0, io_js_1.readChannels)(channels, select);
    }
}
exports._localRead = _localRead;
function _localWrite(
// eslint-disable-next-line @typescript-eslint/no-explicit-any
commit, processes, channels, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
writes) {
    for (const [chan, value] of writes) {
        if (chan === constants_js_1.TASKS) {
            if (!(0, constants_js_1._isSend)(value)) {
                throw new errors_js_1.InvalidUpdateError(`Invalid packet type, expected SendProtocol, got ${JSON.stringify(value)}`);
            }
            if (!(value.node in processes)) {
                throw new errors_js_1.InvalidUpdateError(`Invalid node name ${value.node} in packet`);
            }
        }
        else if (!(chan in channels)) {
            console.warn(`Skipping write for channel '${chan}' which has no readers`);
        }
    }
    commit(writes);
}
exports._localWrite = _localWrite;
function _applyWrites(checkpoint, channels, tasks, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
getNextVersion) {
    // Update seen versions
    for (const task of tasks) {
        if (checkpoint.versions_seen[task.name] === undefined) {
            checkpoint.versions_seen[task.name] = {};
        }
        for (const chan of task.triggers) {
            if (chan in checkpoint.channel_versions) {
                checkpoint.versions_seen[task.name][chan] =
                    checkpoint.channel_versions[chan];
            }
        }
    }
    // Find the highest version of all channels
    let maxVersion;
    if (Object.keys(checkpoint.channel_versions).length > 0) {
        maxVersion = (0, langgraph_checkpoint_1.maxChannelVersion)(...Object.values(checkpoint.channel_versions));
    }
    // Consume all channels that were read
    const channelsToConsume = new Set(tasks
        .flatMap((task) => task.triggers)
        .filter((chan) => !constants_js_1.RESERVED.includes(chan)));
    for (const chan of channelsToConsume) {
        if (channels[chan].consume()) {
            if (getNextVersion !== undefined) {
                checkpoint.channel_versions[chan] = getNextVersion(maxVersion, channels[chan]);
            }
        }
    }
    // Clear pending sends
    if (checkpoint.pending_sends) {
        checkpoint.pending_sends = [];
    }
    // Group writes by channel
    const pendingWriteValuesByChannel = {};
    for (const task of tasks) {
        for (const [chan, val] of task.writes) {
            if (chan === constants_js_1.TASKS) {
                checkpoint.pending_sends.push({
                    node: val.node,
                    args: val.args,
                });
            }
            else {
                if (chan in pendingWriteValuesByChannel) {
                    pendingWriteValuesByChannel[chan].push(val);
                }
                else {
                    pendingWriteValuesByChannel[chan] = [val];
                }
            }
        }
    }
    // find the highest version of all channels
    maxVersion = undefined;
    if (Object.keys(checkpoint.channel_versions).length > 0) {
        maxVersion = (0, langgraph_checkpoint_1.maxChannelVersion)(...Object.values(checkpoint.channel_versions));
    }
    const updatedChannels = new Set();
    // Apply writes to channels
    for (const [chan, vals] of Object.entries(pendingWriteValuesByChannel)) {
        if (chan in channels) {
            let updated;
            try {
                updated = channels[chan].update(vals);
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
            }
            catch (e) {
                if (e.name === errors_js_1.InvalidUpdateError.unminifiable_name) {
                    throw new errors_js_1.InvalidUpdateError(`Invalid update for channel ${chan} with values ${JSON.stringify(vals)}`);
                }
                else {
                    throw e;
                }
            }
            if (updated && getNextVersion !== undefined) {
                checkpoint.channel_versions[chan] = getNextVersion(maxVersion, channels[chan]);
            }
            updatedChannels.add(chan);
        }
    }
    // Channels that weren't updated in this step are notified of a new step
    for (const chan of Object.keys(channels)) {
        if (!updatedChannels.has(chan)) {
            const updated = channels[chan].update([]);
            if (updated && getNextVersion !== undefined) {
                checkpoint.channel_versions[chan] = getNextVersion(maxVersion, channels[chan]);
            }
        }
    }
}
exports._applyWrites = _applyWrites;
function _prepareNextTasks(checkpoint, processes, channels, config, forExecution, extra) {
    const parentNamespace = config.configurable?.checkpoint_ns ?? "";
    const tasks = [];
    const taskDescriptions = [];
    const { step, isResuming = false, checkpointer, manager } = extra;
    for (const packet of checkpoint.pending_sends) {
        if (!(0, constants_js_1._isSendInterface)(packet)) {
            console.warn(`Ignoring invalid packet ${JSON.stringify(packet)} in pending sends.`);
            continue;
        }
        if (!(packet.node in processes)) {
            console.warn(`Ignoring unknown node name ${packet.node} in pending sends.`);
            continue;
        }
        const triggers = [constants_js_1.TASKS];
        const metadata = (0, utils_js_1._getIdMetadata)({
            langgraph_step: step,
            langgraph_node: packet.node,
            langgraph_triggers: triggers,
            langgraph_task_idx: forExecution ? tasks.length : taskDescriptions.length,
        });
        const checkpointNamespace = parentNamespace === ""
            ? packet.node
            : `${parentNamespace}${constants_js_1.CHECKPOINT_NAMESPACE_SEPARATOR}${packet.node}`;
        const taskId = (0, langgraph_checkpoint_1.uuid5)(JSON.stringify([checkpointNamespace, metadata]), checkpoint.id);
        if (forExecution) {
            const proc = processes[packet.node];
            const node = proc.getNode();
            if (node !== undefined) {
                const writes = [];
                tasks.push({
                    name: packet.node,
                    input: packet.args,
                    proc: node,
                    writes,
                    triggers,
                    config: (0, runnables_1.patchConfig)((0, runnables_1.mergeConfigs)(config, processes[packet.node].config, {
                        metadata,
                    }), {
                        runName: packet.node,
                        callbacks: manager?.getChild(`graph:step:${step}`),
                        configurable: {
                            [constants_js_1.CONFIG_KEY_SEND]: _localWrite.bind(undefined, (items) => writes.push(...items), processes, channels),
                            [constants_js_1.CONFIG_KEY_READ]: _localRead.bind(undefined, checkpoint, channels, {
                                name: packet.node,
                                writes: writes,
                                triggers,
                            }),
                        },
                    }),
                    id: taskId,
                    retry_policy: proc.retryPolicy,
                });
            }
        }
        else {
            taskDescriptions.push({ id: taskId, name: packet.node });
        }
    }
    // Check if any processes should be run in next step
    // If so, prepare the values to be passed to them
    const nullVersion = (0, utils_js_1.getNullChannelVersion)(checkpoint.channel_versions);
    if (nullVersion === undefined) {
        return forExecution ? tasks : taskDescriptions;
    }
    for (const [name, proc] of Object.entries(processes)) {
        const seen = checkpoint.versions_seen[name] ?? {};
        const triggers = proc.triggers
            .filter((chan) => {
            const result = (0, io_js_1.readChannel)(channels, chan, false, true);
            const isEmptyChannelError = 
            // eslint-disable-next-line no-instanceof/no-instanceof
            result instanceof Error &&
                result.name === errors_js_1.EmptyChannelError.unminifiable_name;
            return (!isEmptyChannelError &&
                (checkpoint.channel_versions[chan] ?? nullVersion) >
                    (seen[chan] ?? nullVersion));
        })
            .sort();
        // If any of the channels read by this process were updated
        if (triggers.length > 0) {
            const val = _procInput(proc, channels, forExecution);
            if (val === undefined) {
                continue;
            }
            const metadata = (0, utils_js_1._getIdMetadata)({
                langgraph_step: step,
                langgraph_node: name,
                langgraph_triggers: triggers,
                langgraph_task_idx: forExecution
                    ? tasks.length
                    : taskDescriptions.length,
            });
            const checkpointNamespace = parentNamespace === ""
                ? name
                : `${parentNamespace}${constants_js_1.CHECKPOINT_NAMESPACE_SEPARATOR}${name}`;
            const taskId = (0, langgraph_checkpoint_1.uuid5)(JSON.stringify([checkpointNamespace, metadata]), checkpoint.id);
            if (forExecution) {
                const node = proc.getNode();
                if (node !== undefined) {
                    const writes = [];
                    tasks.push({
                        name,
                        input: val,
                        proc: node,
                        writes,
                        triggers,
                        config: (0, runnables_1.patchConfig)((0, runnables_1.mergeConfigs)(config, proc.config, { metadata }), {
                            runName: name,
                            callbacks: manager?.getChild(`graph:step:${step}`),
                            configurable: {
                                [constants_js_1.CONFIG_KEY_SEND]: _localWrite.bind(undefined, (items) => writes.push(...items), processes, channels),
                                [constants_js_1.CONFIG_KEY_READ]: _localRead.bind(undefined, checkpoint, channels, {
                                    name,
                                    writes: writes,
                                    triggers,
                                }),
                                [constants_js_1.CONFIG_KEY_CHECKPOINTER]: checkpointer,
                                [constants_js_1.CONFIG_KEY_RESUMING]: isResuming,
                                checkpoint_id: checkpoint.id,
                                checkpoint_ns: checkpointNamespace,
                            },
                        }),
                        id: taskId,
                        retry_policy: proc.retryPolicy,
                    });
                }
            }
            else {
                taskDescriptions.push({ id: taskId, name });
            }
        }
    }
    return forExecution ? tasks : taskDescriptions;
}
exports._prepareNextTasks = _prepareNextTasks;
function _procInput(proc, channels, forExecution) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let val;
    // If all trigger channels subscribed by this process are not empty
    // then invoke the process with the values of all non-empty channels
    if (Array.isArray(proc.channels)) {
        let successfulRead = false;
        for (const chan of proc.channels) {
            try {
                val = (0, io_js_1.readChannel)(channels, chan, false);
                successfulRead = true;
                break;
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
            }
            catch (e) {
                if (e.name === errors_js_1.EmptyChannelError.unminifiable_name) {
                    continue;
                }
                else {
                    throw e;
                }
            }
        }
        if (!successfulRead) {
            return;
        }
    }
    else if (typeof proc.channels === "object") {
        val = {};
        try {
            for (const [k, chan] of Object.entries(proc.channels)) {
                val[k] = (0, io_js_1.readChannel)(channels, chan, !proc.triggers.includes(chan));
            }
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        }
        catch (e) {
            if (e.name === errors_js_1.EmptyChannelError.unminifiable_name) {
                return;
            }
            else {
                throw e;
            }
        }
    }
    else {
        throw new Error(`Invalid channels type, expected list or dict, got ${proc.channels}`);
    }
    // If the process has a mapper, apply it to the value
    if (forExecution && proc.mapper !== undefined) {
        val = proc.mapper(val);
    }
    return val;
}
